easybuild/easyblocks/g/gromacs.py

```diff
@@ -147,6 +150,33 @@ def prepare_step(self, *args, **kwargs):
     def configure_step(self):
         """Custom configuration procedure for GROMACS: set configure options for configure or cmake."""
 
+        if LooseVersion(self.version) >= LooseVersion('4.6'):
+            cuda = get_software_root('CUDA')
+            if cuda:
+                # CUDA and double precision is currently not supported
+                # If easyconfig explicitly have double_precision=True error out,
+                # otherwise warn about it and skip the double precision build
+                if self.cfg['double_precision']:
+                    raise EasyBuildError("Double precision is not available for GPU build. " +
+                                         "Please explicitly set \"double_precision = False\" " +
+                                         "or remove it in the easyconfig file.")
+                if re.search('DGMX_DOUBLE=(ON|YES|TRUE|Y|[1-9])', self.cfg.get('configopts'), re.I):
+                    if self.cfg['double_precision'] is None:
+                        # Only print warning once when trying double precision
+                        # build the frst time
+                        self.cfg['double_precision'] = False
+                        print_warning("Double precision is not available for " +
```

Comment:
None means that the easyblock decides.

The warning "Double precision is not available for GPU build. Skipping the double precision build." is only printed if this is a CUDA build and -DGMX_DOUBLE=Y is found in the easyconfig and double_precision is unset in the easyconfig.

The "remove it in the easyconfig file." is from the warning before that.
So those warnings are doing the right thing. It's telling you that CUDA and double precision is not available, and that it is skipping it. Unless you have explicitly set double_precision = True, then it is bailing out and telling you it can't go on since double precision and CUDA is not possible.

Reply:
> None means that the easyblock decides.

This should be documented together with how it decides.

> Unfortunately I can't check for CUDA in run_steps...

So that verbose warning will always be printed for default builds? That's not acceptable IMO. As I see the warning is not printed when `self.cfg['double_precision'] == False`, then why not set it to `False` in `run_all_steps`? Also what I don't understand: This now builds single and double precision in one go. So isn't `self.cfg['double_precision']` meaningless? Shouldn't `run_all_steps` take that into account? So `True=>only double`, `False=>Only single`, `None=>Both (with CUDA silently skipped)`? It should still be set to `False` by `run_all_steps` to avoid this warning.

Reply:
In the end I'd prefer to drop the double_precision option. But that's my personal opinion.

Yes, could check if it is set in run_all_steps and only build accordingly, that's a good idea.

But since I can't check which dependencies are available in run_all_steps I need to check it after they become available.
And since it does print that it is going to build double precision (as long as it is set to None) , (the "self.log.debug("List of configure options to iterate over: %s", self.cfg.get('configopts'))" part), even for a CUDA build (since it doesn't know that yet) it's only proper to tell the user that it found out that it can't.

And setting it to False in run_all_steps will make it miss checking for double precision files in sanity_check

Reply:
Oh, and I already do check for double_precision when setting up the build types.

Reply:
> Unfortunately I can't check for CUDA in run_steps...

An idea: Can't you do `isinstance(self.toolchain, Cuda)` with `from easybuild.toolchains.compiler.cuda import Cuda`

Reply:
> > Unfortunately I can't check for CUDA in run_steps...
> 
> An idea: Can't you do `isinstance(self.toolchain, Cuda)` with `from easybuild.toolchains.compiler.cuda import Cuda`

This would introduce the requirement that CUDA is part of the toolchain, rather than potentially just a simple dependency (not everyone is on the `fosscuda` train yet)

Reply:
But it would work for the offical ECs. And thirdparties have another reason to switch to toolchain-enabled CUDA: A fat warning printed by this EB

Reply:
Of course, it is just a warning, sorry for the noise

Reply:
No, don't misunderstand me. It is good you mentioning this. So we know that we could do that proposed check to avoid the `print_warning` being called for official ECs but also need the check for CUDA when non-official ECs are used.

Reply:
The warning is now only in the log. Still haven't found a way to detect CUDA in a convenient place that gets rid of the extra iterations for double precision.